### ISSUE TITLE
Failure to infer type argument for generic function used in another generic function

### ISSUE THREAD

**Issue by pkch** [python/mypy#3146]
In this example, it seems mypy has enough information to infer that `compare()` should be used with type argument set to `str`, so this should type check:

```
T = TypeVar('T')

class X(Generic[T]): ...

def compare(x: X[T], y: X[T]) -> None: ...

def factory() -> X[T]: return X[T]()

y: X[str]
compare(factory(), y)  # Cannot infer type argument 1 of "compare"
```

With a slight modification, the type error mysteriously disappears. However, the result is incorrect: `Any` is revealed instead of `str`.

```
T = TypeVar('T')

class X(Generic[T]): ...

def compare(x: X[T], y: X[T]) -> T: ...

def factory() -> X[T]: return X[T]()

y: X[str]
z = compare(X[T](), y)
reveal_type(z)  # Any
```

**Comment by gvanrossum** [comment]
This is a duplicate, it's because factory() returns a generic function but is itself not generic. @soxolet is already working on a solution.

**Comment by ilevkivskyi** [comment]
There is something more, this line ``z = compare(X[T](), y)`` should be flagged as error, since ``T`` is unbound here.

**Comment by gvanrossum** [comment]
Yeah, but I'm sure we already have a bug for that too...
